File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.cpp
class TrieNode{
    
    public:
        TrieNode* next[26]={NULL};
        bool flag=false;
};



class WordDictionary {
public:
     TrieNode* root=NULL;
    WordDictionary() {
        
        root=new TrieNode();
        
        
    }
    
    void addWord(string word) {
        
        TrieNode* temp=root;
        for(auto w : word){
            
                if(temp->next[w-'a']==NULL){
                    
                            temp->next[w-'a']=new TrieNode();
                    
                }
                temp=temp->next[w-'a'];
        }
       
        temp->flag=true;
        
    }
    
    bool dfs(TrieNode* root,string& word,int indx){
        
            if(root==NULL){
                    return false;
            }
        
             if(indx==(word.size())){
                
                    if(word[indx]=='.'){
                        
                        for(int i=0;i<26;i++){
                            if(root->next[i]!=NULL){
                                   return true;
                            }   
                         }            
                    }
                //cout<<root->flag<<"\t"<<word<<indx<<endl;
                return root->flag;
            }
    
            if(word[indx]=='.'){
                    bool val=false;
                    for(int i=0;i<26;i++){
                            if(root->next[i]!=NULL){
                                   val|=dfs(root->next[i],word,indx+1);
                            }   
                    }
                    return val;
            }
        
            return dfs(root->next[word[indx]-'a'],word,indx+1); 
        
    }
    
    
    
    bool search(string word) {
        
           TrieNode* temp=root;
           return dfs(temp,word,0);
        
        
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class TrieNode:
    
    def __init__(self):
        self.child=[None]*26
        self.flag=False



class WordDictionary:

    def __init__(self):
        
        self.root=TrieNode()
        

    def addWord(self, word: str) -> None:
        
        temp=self.root
        for char in word:
            val=ord(char)-97
            if temp.child[val]==None:
                temp.child[val]=TrieNode()
            temp=temp.child[val]
            
    
        temp.flag=True
        
    def dfs(self,temp,word,indx):
        
        if temp==None:
                return False
        
        if indx==len(word):
            print(temp.flag)
            return temp.flag
        
        val=False
        if word[indx]=='.':
                for item in temp.child:
                        if temp!=None:
                                val |=self.dfs(item,word,indx+1)
        else:
                print(word[indx])
                if temp.child[ord(word[indx])-97]==None:
                        return False
                    
                temp=temp.child[ord(word[indx])-97]
                val|=self.dfs(temp,word,indx+1)    
                
                
                
        return val
    
    def search(self, word: str) -> bool:
        
        temp=self.root
        return self.dfs(temp,word,0)
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)